src/custom-elements/src/components/number-input/number-input.tsx

```diff
@@ -0,0 +1,182 @@
+import { CallbackQueue } from '../../utils/CallbackQueue'
+
+@Component({
+  tag: 'chi-number-input',
+  styleUrl: 'number-input.scss',
+  scoped: true
+})
+export class NumberInput {
+
+  private _didUpdateCallBackOnceQueue: (() => void)[] = [];
+
+  private initialValue;
+
+  /**
+   * used to hold the value of the number input
+   */
+  @Prop({ mutable: true, reflectToAttr: true }) value: string = '0';
+
+  /**
+   * used to set a step to indicate the expected granularity
+   */
+  @Prop({ reflectToAttr: true }) step: number = 1;
+
+  /**
+   * used to set a maximum allowed value
+   */
+  @Prop({ reflectToAttr: true }) max: number = Number.MAX_SAFE_INTEGER;
+
+  /**
+   * used to set a minimum allowed value
+   */
+  @Prop({ reflectToAttr: true }) min: number = Number.MIN_SAFE_INTEGER;
+
+  /**
+   * used to determine if component is disabled or not
+   */
+  @Prop({ reflectToAttr: true }) disabled?: boolean;
+
+  /**
+   * used to determine component's size
+   */
+  @Prop({ reflectToAttr: true }) size?: string;
+
+  /**
+   * used to set component's expanded mode
+   */
+  @Prop({ reflectToAttr: true }) expanded?: boolean;
+
+  /**
+   * used to determine if expanded component has a pill look
+   */
+  @Prop({ reflectToAttr: true }) pill?: boolean;
+
+  /**
+   * used to provide an input style like 'danger'. Mostly used for testing purposes
+   */
+  @Prop() inputstyle?: string;
+
+  /**
+   * used to provide an input state like 'hover' or 'focus'. Mostly used for testing purposes
+   */
+  @Prop() state?: string;
+
+  @Event() chiChange: EventEmitter<string>;
+
+  componentDidLoad() {
+    this.initialValue = this.value;
+  }
+
+  componentDidUpdate() {
+    CallbackQueue.queueProcess(this._didUpdateCallBackOnceQueue);
+  }
+
+  handleChange(ev: Event) {
+    let stepDifference = 0;
+
+    this.value = !!ev.target ? (ev.target as HTMLInputElement).value.toString() : null;
+
+    stepDifference = Math.round((+this.value % this.step - this.initialValue) * 10000) / 10000;
+
+    if (stepDifference !== 0) {
+      this.value = (Math.round(((+this.value - stepDifference) * 10000) / 10000) + this.step).toString();
+    }
+
+    if ((+this.value > this.max) || (+this.value < this.min)) {
+      stepDifference = Math.round((+this.value % this.step - this.initialValue) * 10000) / 10000;
+
+      if (stepDifference !== 0) {
+        this.value = (Math.round(((+this.value - stepDifference) * 10000) / 10000) + this.step).toString();
+      }
+    }
+    
+    this._didUpdateCallBackOnceQueue.push(()=>{
+      this.chiChange.emit(this.value);
+    });
+  }
+
+  private increment() {
+    const step = this.step;
+    const newValue = Math.round((+this.value + step) * 10000) / 10000;
+
+    if (newValue > this.max) {
+      return;
+    } else {
+      this.value =  newValue.toString();
+      this._didUpdateCallBackOnceQueue.push(()=>{
+        this.chiChange.emit(this.value);
+      });
+    }
```

Comment:
Is the queue really needed why can we just emit the event directly? I guess I'm may be missing some benefit here.

Reply:
We must raise the event when we are sure the value has really changed, that's why emits are enqueued. I think @dani-cl-madrid can provide more detailed information about this.

Reply:
can we replace if else with just if.

`
if (newValue < this.max) {
	      this.value =  newValue.toString();
	      this._didUpdateCallBackOnceQueue.push(() => {
	        this.chiChange.emit(this.value);
	      });
	    }
`